runtime/registry.go

```diff
@@ -0,0 +1,290 @@
+	// registered for a key that matches keyOrPrefix.
+	if isPrefixKey(keyOrPrefix) {
+		foundProvider := ""
+		r.providers.WalkPrefix(keyOrPrefix, func(s string, _ interface{}) bool {
+			foundProvider = s
+			return true
+		})
+		if foundProvider != "" {
+			return nil, fmt.Errorf("%w: found provider on %s", ErrKeyTaken, key)
+		}
+	}
+
+	r.providers.Insert(keyOrPrefix, p)
+
+	return func(records ...record.Record) {
+		r.l.RLock()
+		defer r.l.RUnlock()
+
+		if r.dbController == nil {
+			return
+		}
+
+		for _, rec := range records {
+			r.dbController.PushUpdate(rec)
+		}
+	}, nil
+}
+
+// Get returns the runtime value that is identified by key.
+// It implements the storage.Interface.
+func (r *Registry) Get(key string) (record.Record, error) {
+	provider := r.getMatchingProvider(key)
+	if provider == nil {
+		return nil, nil
+	}
+
+	records, err := provider.Get(key)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get performs an exact match so filter out
+	// and values that do not match key.
+	for _, r := range records {
+		if r.DatabaseKey() == key {
+			return r, nil
+		}
+	}
+	return nil, nil
+}
+
+// Put stores the record m in the runtime database. Note that
+// ErrReadOnly is returned if there's no value provider responsible
+// for m.Key().
+func (r *Registry) Put(m record.Record) (record.Record, error) {
+	provider := r.getMatchingProvider(m.DatabaseKey())
+	if provider == nil {
+		// if there's no provider for the given value
+		// return ErrKeyUnmanaged.
+		// TODO(ppacher): we could also create a memoryProvider that
+		// just stores the record in memory. Though, not sure we need
+		// that.
+		return nil, ErrKeyUnmanaged
+	}
+
+	res, err := provider.Set(m)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Query performs a query on the runtime registry returning all
+// records across all value providers that match q.
+// Query implements the storage.Storage interface.
+func (r *Registry) Query(q *query.Query, local, internal bool) (*iterator.Iterator, error) {
+	if _, err := q.Check(); err != nil {
+		return nil, fmt.Errorf("invalid query: %w", err)
+	}
+
+	searchPrefix := q.DatabaseKeyPrefix()
+	providers := r.collectProviderByPrefix(searchPrefix)
+	if len(providers) == 0 {
+		return nil, fmt.Errorf("%w: for key %s", ErrKeyUnmanaged, searchPrefix)
+	}
+
+	iter := iterator.New()
+
+	grp := new(errgroup.Group)
+	for idx := range providers {
+		p := providers[idx]
+
+		grp.Go(func() error {
+			key := p.key
+			if len(searchPrefix) > len(key) {
+				key = searchPrefix
+			}
+
+			records, err := p.Get(key)
+			if err != nil {
+				return err
+			}
+
+			for _, r := range records {
+				// TODO(ppacher): do we need to lock r?
+				//                storage/hashmap does not lock the records
+				//                before sending them to the iterator but
+				//                better make sure that's correct.
+
+				if !strings.HasPrefix(r.DatabaseKey(), searchPrefix) {
+					continue
+				}
```

Comment:
I wanted to keep the logic on the `ValueProvider` side as simple as possible. So a dumb value provider can actually ignore the `key` parameter in `Get()` and return all records it provides. The prefix check makes sure we quickly filter the responses of each value provider. Since there won't be that much and queries only come from the API side there won't be any performance issues if we would check those records twice.
Still, we need to pass the `key` to `Get()` because if the provider does a lot or is mounted/registered at different keys those parameter can become very handy.